node/network/collator-protocol/src/validator_side.rs

```diff
@@ -590,23 +676,50 @@ where
 		..Default::default()
 	};
 
+	let next_inactivity_stream = futures::stream::unfold(
+		Instant::now() + ACTIVITY_POLL,
+		|next_check| async move { Some(((), wait_until_next_check(next_check).await)) }
+	);
+
+	futures::pin_mut!(next_inactivity_stream);
+
 	loop {
-		if let Poll::Ready(msg) = futures::poll!(ctx.recv()) {
-			let msg = msg?;
-			tracing::trace!(target: LOG_TARGET, msg = ?msg, "received a message");
-
-			match msg {
-				Communication { msg } => process_msg(&mut ctx, msg, &mut state).await,
-				Signal(BlockFinalized(..)) => {}
-				Signal(ActiveLeaves(_)) => {}
-				Signal(Conclude) => { break }
+		let res = {
+			let s = futures::future::select(ctx.recv().fuse(), next_inactivity_stream.next().fuse());
```

Comment:
Why don't you use the `select` macro? This code looks more complicated than it needs to.

And you should also fuse the stream outside the loop, otherwise the fuse doesn't make that much sense.

Reply:
This is different than the `select!` macro because it does not poll the future to completion but instead polls it just a single time. This is because of the `collation_requests` map polling logic below that. That logic should be refactored to use `FuturesUnordered` but I spent about an hour trying to do that refactoring and it is not that simple.

Reply:
Yeah sorry I had not seen this logic below. Yeah that complicates the stuff :(